src/go/concurr.go

```diff
@@ -21,27 +23,21 @@ var	bucket = os.Args[1]
 var	input_file = os.Args[2]
 var	output_dir = os.Args[3]
 
-var files []string
-
-func worker(downloader *s3manager.Downloader, wg *sync.WaitGroup) {
-	var filename string
-	for len(files) != 0 {
-		// Pop from the front, (or shift)
-		filename, files = files[0], files[1:]
+var files chan string
 
+func worker(downloader *s3manager.Downloader) {
+	for filename := range files {
 		outfile, err := os.Create(filepath.Join(output_dir, filename))
 		shared.Check(err)
+		defer outfile.Close()
```

Comment:
yup this was me being dumb. I'd prefer to defer to assure the file is closed correctly, but the defer will run at the end of the function, so each iteration of the loop would have to be it's own function to use defer here, otherwise we need to do the manual close and just assure it happens in all cases